File: stream-client.js
const fs = require('fs-extra');
const io = require('socket.io-client')('http://localhost:8000');
const sha1File = require('sha1-file');

io.on('connect', async () => {

  var prop = {};
  // prop['inFile'] = 'in.png';
  // prop['inFile'] = '1G.iso';
  prop['inFile'] = 'c.jpg'; // 5MB file
  var byteCount = 65 * 1024;
  var f = await fs.stat(prop['inFile'])
  prop['size'] = f.size;
  prop['ext'] = prop['inFile'].split('.').pop();
  var sha1 = await sha1File(prop['inFile']);

  prop['sha1'] = sha1;
  prop['start'] = 0;

  console.log(prop);
  io.emit('UPLOAD_START', prop);

  var stream;

  var start = Date.now();

  io.on('UPLOAD_START', (data) => {
    prop['filename'] = data.filename;
    prop['start'] = data.start;
    console.log(data);
    stream = fs.createReadStream(prop['inFile'], {start: prop['start'], highWaterMark: byteCount});    
    stream.on('readable', () => {
      prop['chunk'] = stream.read();
      io.emit('UPLOAD', prop);
      try {
        if (prop['chunk'] == null) {
            // console.log('Done UPLOAD');
            return;
        }
//        console.log(prop['chunk'].slice(0,5));
        prop['start'] = prop['start'] + prop['chunk'].length;
      } catch (e) {
        console.log(e);
      }
      console.log('Progress: ' + prop['start'] / prop['size'] * 100);
    });

    stream.on('end', () => {
      // console.log('STREAM ENDED');
      io.emit('UPLOAD_END');
    });
  });

  io.on('UPLOAD_END', (data) => {
    console.log('UPLOAD_END');
//    console.log('SHA1: ' + prop['sha1']);
    console.log('');
    console.log('Language: JAVSCRIPT');
    console.log('Stream block size: 65 * 1024');
    console.log('Size in bytes: ' + prop['size']);
    var end = Date.now();
    var elapsed = end - start;
    console.log('Elapsed in ms: ' + elapsed);
  });


});

File: stream-client.dart
import 'package:socket_io_client/socket_io_client.dart' as IO;
import 'dart:io';
import 'dart:convert';
import 'package:convert/convert.dart';

main() async {
  var socket = IO.io('http://localhost:8000', <String, dynamic>{
    'transports': ['websocket'],
    'upgrade': false
  });
  

  // var file = 'in.png';
  // var file = '1G.iso';
  var file = 'c.jpg'; // 5MB file
  var f = new File(file);
  Map prop = {};
  prop['ext'] = file.split('.').last;
  prop['size'] = await f.length();
  prop['start'] = 0;
  print(prop);

  var start = new DateTime.now().millisecondsSinceEpoch;

socket.on('connect', (s) {
  var stream = f.openRead();
  var length = 0;

  socket.emit('UPLOAD_START', prop);
  stream
    // .transform(Uint8List.buffer)
    .listen( (data) {
    // print('Sending ...');
    prop['chunk'] = data;
    prop['start'] = prop['start'] + data.length;
    print('Progress: ' + (prop['start'] / prop['size'] * 100).toString() );
    socket.emitWithBinary('UPLOAD_DART', prop);
  }, onDone: () => socket.emit('UPLOAD_END'));

});

  socket.on('UPLOAD_END', (data) {   
    print('UPLOAD_END');
    var end = new DateTime.now().millisecondsSinceEpoch;
    var elapsed = end - start;
    print('');
    print('Language: DART');
    print('Stream block size: 65 * 1024');
    print('Size in bytes: ' + prop['size'].toString());
    print('Elasped in ms:  ' + elapsed.toString());
  });

}
